src/optimizer/rule/PushFilterDownAggregateRule.cpp

```diff
@@ -0,0 +1,127 @@
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "optimizer/rule/PushFilterDownAggregateRule.h"
+
+#include "optimizer/OptContext.h"
+#include "optimizer/OptGroup.h"
+#include "planner/PlanNode.h"
+#include "planner/Query.h"
+#include "util/ExpressionUtils.h"
+#include "visitor/ExtractFilterExprVisitor.h"
+#include "visitor/RewriteVisitor.h"
+
+using nebula::graph::PlanNode;
+using nebula::graph::QueryContext;
+
+namespace nebula {
+namespace opt {
+
+std::unique_ptr<OptRule> PushFilterDownAggregateRule::kInstance =
+    std::unique_ptr<PushFilterDownAggregateRule>(new PushFilterDownAggregateRule());
+
+PushFilterDownAggregateRule::PushFilterDownAggregateRule() {
+    RuleSet::QueryRules().addRule(this);
+}
+
+const Pattern& PushFilterDownAggregateRule::pattern() const {
+    static Pattern pattern = Pattern::create(graph::PlanNode::Kind::kFilter,
+                                             {Pattern::create(graph::PlanNode::Kind::kAggregate)});
+    return pattern;
+}
+
+StatusOr<OptRule::TransformResult> PushFilterDownAggregateRule::transform(
+    OptContext* octx,
+    const MatchedResult& matched) const {
+    auto qctx = octx->qctx();
+    auto* filterGroupNode = matched.node;
+    auto* oldFilterNode = filterGroupNode->node();
+    auto deps = matched.dependencies;
+    DCHECK_EQ(deps.size(), 1);
+    auto aggGroupNode = deps.front().node;
+    auto* oldAggNode = aggGroupNode->node();
+    DCHECK(oldFilterNode->kind() == PlanNode::Kind::kFilter);
+    DCHECK(oldAggNode->kind() == PlanNode::Kind::kAggregate);
+    auto* newFilterNode =
+        const_cast<graph::Filter*>(static_cast<const graph::Filter*>(oldFilterNode))->clone(qctx);
+    auto* newAggNode =
+        const_cast<graph::Aggregate*>(static_cast<const graph::Aggregate*>(oldAggNode))
+            ->clone(qctx);
+    const auto* condition = newFilterNode->condition();
+    auto& groupItems = newAggNode->groupItems();
+
+    // Check expression recursively to ensure no aggregate items in the filter
+    auto varProps = graph::ExpressionUtils::collectAll(condition, {Expression::Kind::kVarProperty});
+    std::vector<std::string> propNames;
+    for (auto* expr : varProps) {
+        DCHECK(expr->kind() == Expression::Kind::kVarProperty);
+        propNames.emplace_back(*static_cast<const VariablePropertyExpression*>(expr)->prop());
+    }
+    std::unordered_map<std::string, Expression*> rewriteMap;
+    auto colNames = newAggNode->colNames();
+    for (size_t i = 0; i < colNames.size(); ++i) {
+        auto& colName = colNames[i];
+        for (size_t j = 0; j < propNames.size(); ++j) {
+            if (!colName.compare(propNames[j])) {
+                if (graph::ExpressionUtils::findAny(groupItems[i],
+                                                    {Expression::Kind::kAggregate})) {
+                    return TransformResult::noTransform();
+                } else {
+                    rewriteMap[colName] = groupItems[i];
```

Comment:
break;

Reply:
plz rethink my former suggestion. I really advise you to refactor the logics with `std::find_if` such as:

```cpp
    for (size_t i = 0; i < colNames.size(); ++i) {
        auto& colName = colNames[i];
        auto iter = std::find_if(propNames.begin(), propNames.end(), [&](const auto& name) {
            return !colName.compare(name);
        });
        if (iter == propNames.end()) continue;
        if (graph::ExpressionUtils::findAny(groupItems[i], {Expression::Kind::kAggregate})) {
            return TransformResult::noTransform();
        }
        rewriteMap[colName] = groupItems[i];
    }
```

It's more clear without many nested blocks.

Reply:
Modified.